File: daily/Linked List/142-Linked-list-cycle-II.py
class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # visitNode = set()
        # while head:
        #     if head in visitNode:
        #         return head
        #     visitNode.add(head)
        #     head = head.next
        # return None
        fast = slow = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                break
        else:
            return None
        
        fast = head
        while fast != slow:
            fast = fast.next
            slow = slow.next
        return fast

File: daily/Linked List/142-Linked-list-cycle-II.go
package daily 

func detectCycle(head *ListNode) *ListNode {
    fast := head
    slow := head
    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
        if slow == fast {
            break
        }
    }
    if fast == nil || fast.Next == nil {
        return nil
    }
    fast = head
    for fast != slow {
        fast = fast.Next
        slow = slow.Next
    }
    return fast
}
